src/mdns/mdns_avahi.cpp

```diff
@@ -1077,33 +1091,43 @@ void PublisherAvahi::ServiceSubscription::HandleResolveResult(AvahiServiceResolv
     VerifyOrExit(!address.IsLinkLocal() && !address.IsMulticast() && !address.IsLoopback() && !address.IsUnspecified(),
                  otbrLogDebug("ignoring address %s", address.ToString().c_str()));
 
-    mInstanceInfo.mAddresses.push_back(address);
+    instanceInfo.mAddresses.push_back(address);
 
     // TODO priority
     // TODO weight
     // TODO use a more proper TTL
-    mInstanceInfo.mTtl = kDefaultTtl;
+    instanceInfo.mTtl = kDefaultTtl;
     for (auto p = aTxt; p; p = avahi_string_list_get_next(p))
     {
         totalTxtSize += avahi_string_list_get_size(p) + 1;
     }
-    mInstanceInfo.mTxtData.resize(totalTxtSize);
-    avahi_string_list_serialize(aTxt, mInstanceInfo.mTxtData.data(), totalTxtSize);
+    instanceInfo.mTxtData.resize(totalTxtSize);
+    avahi_string_list_serialize(aTxt, instanceInfo.mTxtData.data(), totalTxtSize);
 
-    otbrLogDebug("resolve service reply: address=%s, ttl=%u", address.ToString().c_str(), mInstanceInfo.mTtl);
+    otbrLogDebug("resolve service reply: address=%s, ttl=%u", address.ToString().c_str(), instanceInfo.mTtl);
 
-    mPublisherAvahi->OnServiceResolved(mType, mInstanceInfo);
+    mPublisherAvahi->OnServiceResolved(mType, instanceInfo);
 
 exit:
     if (avahi_client_errno(mPublisherAvahi->mClient) != AVAHI_OK)
     {
         mPublisherAvahi->OnServiceResolveFailed(*this, avahi_client_errno(mPublisherAvahi->mClient));
     }
-    if (mServiceBrowser != nullptr)
-    {
-        avahi_service_resolver_free(mServiceResolver);
-        mServiceResolver = nullptr;
-    }
+
+    RemoveServiceResolver(aServiceResolver);
+}
+
+void PublisherAvahi::ServiceSubscription::AddServiceResolver(AvahiServiceResolver *aServiceResolver)
+{
+    assert(aServiceResolver != nullptr);
+    mServiceResolvers.insert(aServiceResolver);
+}
+
+void PublisherAvahi::ServiceSubscription::RemoveServiceResolver(AvahiServiceResolver *aServiceResolver)
+{
+    assert(aServiceResolver != nullptr);
+    avahi_service_resolver_free(aServiceResolver);
+    mServiceResolvers.erase(aServiceResolver);
```

Comment:
I'd suggest to move this line before `avahi_service_resolver_free`. There is no bug with current code, but if `avahi_service_resolver_free` uses `T *&` as the parameter type this is dangerous to consider `aServiceResolver` as  valid after freeing it.

Reply:
SGTM.